core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java

```diff
@@ -306,12 +307,30 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                                                         .contentLength(req.headers())
                                                         .transferred(transferredLength)
                                                         .build();
-                        fail(id, decodedReq.headers(), HttpStatus.REQUEST_ENTITY_TOO_LARGE, null, cause);
+                        discarding = true;
+                        req = null;
+                        final boolean shouldReset;
+                        if (encoder instanceof ServerHttp1ObjectEncoder) {
+                            keepAliveHandler.disconnectWhenFinished();
+                            shouldReset = false;
+                        } else {
+                            // Upgraded to HTTP/2. Reset only if the remote peer is still open.
+                            shouldReset = !endOfStream;
+                        }
+
                         // Wrap the cause with the returned status to let LoggingService correctly log the
                         // status.
-                        decodedReq.abortResponse(
-                                HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause),
-                                true);
+                        final HttpStatusException httpStatusException =
+                                HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause);
+                        if (!decodedReq.isInitialized()) {
+                            assert decodedReq.needsAggregation();
+                            final StreamingDecodedHttpRequest streamingReq = decodedReq.toAbortedStreaming(
+                                    inboundTrafficController, httpStatusException, shouldReset);
+                            ctx.fireChannelRead(streamingReq);
```

Comment:
Question) It seems like we don't early return when `encoder` is swapped to http2 and an upgrade request is received.

Is there a code point where we guard against `ctx.fireChannelRead` being invoked twice? (once for HTTP1 upgrade, once for HTTP2)

Reply:
Sure. 😉 
An `UpgradeEvent` is created when it receives the upgrade event:
https://github.com/line/armeria/blob/b7c14c4564f7137b421292fb01bda0ee58fbe93b/core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java#L297
`Http2ConnectionHandler` is added to the pipleline:
https://github.com/line/armeria/blob/e7be74e738a49448afe98773e1bbf7ab09080c2f/core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java#L109
`Http1RequestDecoder` receives the `UpgradeEvent` and calls `channelRead()` with the HTTP/1.1 request:
https://github.com/line/armeria/blob/091d193bac243d1463a03bd42a5174b88adb479e/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L453
After handling the request, `Http1RequestDecoder` is removed:
https://github.com/line/armeria/blob/091d193bac243d1463a03bd42a5174b88adb479e/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L275-L279

Reply:
My question was rather when `ctx.pipeline().remove(this); ` is called, but the content length of the request is exceeded.

Let me do some testing in my local env. instead and follow up if needed, feel free to merge 😄

Reply:
I misunderstood it. 😓 
No, I didn't consider that case and I have to fix it. Thanks!